packages/integrations/sitemap/src/index.ts

```diff
@@ -51,18 +54,71 @@ const OUTFILE = 'sitemap-index.xml';
 
 const createPlugin = (options?: SitemapOptions): AstroIntegration => {
 	let config: AstroConfig;
+	const logger = new Logger(PKG_NAME);
+
 	return {
 		name: PKG_NAME,
 
 		hooks: {
 			'astro:config:done': async ({ config: cfg }) => {
-				config = cfg;
+				if (cfg.site) {
+					config = cfg;
+				} else {
+					// eslint-disable-next-line no-console
+					console.warn(
+						'The Sitemap integration requires the `site` astro.config option. Skipping.'
+					);
+					return;
+				}
 			},
 
-			'astro:build:done': async ({ dir, pages }) => {
-				const logger = new Logger(PKG_NAME);
+			'astro:build:start': async () => {
+				if (config.output !== 'server' || !config.site) {
+					return;
+				}
+
+				const srcPath = fileURLToPath(config.srcDir);
+				const pagesPath = resolve(srcPath, 'pages');
+
+				const pageFiles = await fg(`${pagesPath}/**/*.{astro,ts,js}`);
+
+				const routes = (
+					await Promise.all(
+						pageFiles.map(async (filePath) => {
+							const isPrerendered = await isRoutePrerendered(filePath);
+							const index = filePath.indexOf('pages/') + 6;
+							const routeSegment = filePath
+								.substring(index)
+								.replace(/\.(astro|ts|js)/, '')
+								.replace(/index$/, '');
+
+							/**
+							 * @TODO
+							 * figure out how to run `getStaticPaths` here.
+							 */
+							const isDynamicRoute = routeSegment.endsWith(']');
+							const shouldIndex = !isDynamicRoute && !isPrerendered;
+
+							return shouldIndex ? `${config.site}/${routeSegment}` : undefined;
+						})
+					)
+				).filter((route): route is string => Boolean(route));
+				const opts = validateOptions(config.site, options);
+
+				opts.customPages = opts.customPages
+					? Array.from(new Set([...routes, ...opts.customPages]))
+					: routes;
+				options = opts;
+
+				logger.info(`build is starting + ${JSON.stringify(opts.customPages, null, 2)}`);
+			},
 
+			'astro:build:done': async ({ dir, pages }) => {
```

Comment:
```suggestion
			'astro:build:done': async ({ dir, routes }) => {
```

Check out the `routes` object which has our full routing manifest. This might have the info you need!